File: typespec/admin/hub-signup-domains/domains.ts
import type { PageSize, PaginationKey } from "../../common/pagination.ts";
import { isPageSize, isPaginationKey } from "../../common/pagination.ts";
import type { HubSignupDomainID } from "../types.ts";
import { isHubSignupDomainID } from "../types.ts";

export type DomainName = string;
export type State = "active" | "disabled";
export type DomainFilterText = string;
export type DisableComment = string;

export const Active: State = "active";
export const Disabled: State = "disabled";

export interface Domain {
  hub_signup_domain_id: HubSignupDomainID;
  domain: DomainName;
  state: State;
  disabled_comment?: DisableComment;
  created_at: string;
  updated_at: string;
}

export interface ListRequest {
  limit?: PageSize;
  pagination_key?: PaginationKey;
  filter_search?: DomainFilterText;
  filter_state?: State;
}

export function normalizeListRequest(request: ListRequest): ListRequest {
  if (request.filter_search === undefined) return { ...request };
  return {
    ...request,
    filter_search: request.filter_search.trim().toLowerCase(),
  };
}

export function effectiveListLimit(request: ListRequest): PageSize {
  return request.limit ?? 50;
}

export function validateListRequest(request: ListRequest): string[] {
  const fields: string[] = [];
  if (!isPageSize(effectiveListLimit(request))) fields.push("limit");
  if (
    request.pagination_key !== undefined &&
    !isPaginationKey(request.pagination_key)
  ) {
    fields.push("pagination_key");
  }
  if (
    request.filter_search !== undefined &&
    !isDomainFilterText(request.filter_search)
  ) {
    fields.push("filter_search");
  }
  if (request.filter_state !== undefined && !isState(request.filter_state)) {
    fields.push("filter_state");
  }
  return fields;
}

export interface ListResponse {
  domains: Domain[];
  next_pagination_key?: PaginationKey;
}

export interface CreateRequest {
  domain: DomainName;
  state?: State;
  disabled_comment?: DisableComment;
}

export function normalizeCreateRequest(request: CreateRequest): CreateRequest {
  if (request.disabled_comment === undefined) {
    return { ...request, domain: normalizeDomainName(request.domain) };
  }
  return {
    ...request,
    domain: normalizeDomainName(request.domain),
    disabled_comment: normalizeDisableComment(request.disabled_comment),
  };
}

export function effectiveCreateState(request: CreateRequest): State {
  return request.state ?? Active;
}

export function validateCreateRequest(request: CreateRequest): string[] {
  const fields: string[] = [];
  if (!isDomainName(request.domain)) fields.push("domain");
  if (!isState(effectiveCreateState(request))) fields.push("state");
  if (
    !isDisableCommentForState(
      effectiveCreateState(request),
      request.disabled_comment,
    )
  ) {
    fields.push("disabled_comment");
  }
  return fields;
}

export interface UpdateRequest {
  hub_signup_domain_id: HubSignupDomainID;
  domain: DomainName;
  state: State;
  disabled_comment?: DisableComment;
}

export function normalizeUpdateRequest(request: UpdateRequest): UpdateRequest {
  if (request.disabled_comment === undefined) {
    return { ...request, domain: normalizeDomainName(request.domain) };
  }
  return {
    ...request,
    domain: normalizeDomainName(request.domain),
    disabled_comment: normalizeDisableComment(request.disabled_comment),
  };
}

export function validateUpdateRequest(request: UpdateRequest): string[] {
  const fields: string[] = [];
  if (!isHubSignupDomainID(request.hub_signup_domain_id)) {
    fields.push("hub_signup_domain_id");
  }
  if (!isDomainName(request.domain)) fields.push("domain");
  if (!isState(request.state)) fields.push("state");
  if (!isDisableCommentForState(request.state, request.disabled_comment)) {
    fields.push("disabled_comment");
  }
  return fields;
}

export function normalizeDomainName(value: DomainName): DomainName {
  return value.trim().toLowerCase().replace(/\.$/, "");
}

export function isDomainName(value: DomainName): boolean {
  const normalized = normalizeDomainName(value);
  if (
    normalized.length < 3 ||
    normalized.length > 253 ||
    !normalized.includes(".") ||
    /^[0-9.]+$/.test(normalized) ||
    !/[a-z]/.test(normalized.split(".").at(-1) ?? "")
  ) {
    return false;
  }
  return normalized
    .split(".")
    .every((label) => /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/.test(label));
}

export function isState(value: State): boolean {
  return value === Active || value === Disabled;
}

export function isDisableComment(value: DisableComment): boolean {
  const length = [...value.trim()].length;
  return length >= 1 && length <= 500;
}

function normalizeDisableComment(value: DisableComment): DisableComment {
  return value.trim();
}

function isDisableCommentForState(
  state: State,
  comment: DisableComment | undefined,
): boolean {
  return state === Disabled
    ? comment !== undefined && isDisableComment(comment)
    : comment === undefined;
}

function isDomainFilterText(value: DomainFilterText): boolean {
  const normalized = value.trim().toLowerCase();
  return (
    normalized.length >= 1 &&
    normalized.length <= 253 &&
    /^[a-z0-9.-]+$/.test(normalized)
  );
}

File: typespec/admin/hub-signup-domains/domains.go
package hubsignupdomains

import (
	"regexp"
	"strings"
	"time"

	adminspec "github.com/vetchium/src/typespec/admin"
	"github.com/vetchium/src/typespec/common"
)

type DomainName string
type State string
type DomainFilterText string
type DisableComment string

const (
	Active   State = "active"
	Disabled State = "disabled"
)

var domainLabelPattern = regexp.MustCompile(
	`^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$`,
)
var numericDomainPattern = regexp.MustCompile(`^[0-9.]+$`)
var topLevelLetterPattern = regexp.MustCompile(`[a-z]`)
var domainFilterPattern = regexp.MustCompile(`^[a-z0-9.-]+$`)

type Domain struct {
	HubSignupDomainID adminspec.HubSignupDomainID `json:"hub_signup_domain_id"`
	Domain            DomainName                  `json:"domain"`
	State             State                       `json:"state"`
	DisabledComment   *DisableComment             `json:"disabled_comment,omitempty"`
	CreatedAt         time.Time                   `json:"created_at"`
	UpdatedAt         time.Time                   `json:"updated_at"`
}

type ListRequest struct {
	Limit         *common.PageSize      `json:"limit,omitempty"`
	PaginationKey *common.PaginationKey `json:"pagination_key,omitempty"`
	FilterSearch  *DomainFilterText     `json:"filter_search,omitempty"`
	FilterState   *State                `json:"filter_state,omitempty"`
}

func (r *ListRequest) Normalize() {
	if r.FilterSearch != nil {
		value := DomainFilterText(strings.ToLower(strings.TrimSpace(
			string(*r.FilterSearch),
		)))
		r.FilterSearch = &value
	}
}

func (r ListRequest) EffectiveLimit() common.PageSize {
	if r.Limit == nil {
		return 50
	}
	return *r.Limit
}

func (r ListRequest) Validate() []string {
	fields := make([]string, 0, 4)
	if !common.IsPageSize(r.EffectiveLimit()) {
		fields = append(fields, "limit")
	}
	if r.PaginationKey != nil && !common.IsPaginationKey(*r.PaginationKey) {
		fields = append(fields, "pagination_key")
	}
	if r.FilterSearch != nil && !isDomainFilterText(*r.FilterSearch) {
		fields = append(fields, "filter_search")
	}
	if r.FilterState != nil && !IsState(*r.FilterState) {
		fields = append(fields, "filter_state")
	}
	return fields
}

type ListResponse struct {
	Domains           []Domain              `json:"domains"`
	NextPaginationKey *common.PaginationKey `json:"next_pagination_key,omitempty"`
}

type CreateRequest struct {
	Domain          DomainName      `json:"domain"`
	State           *State          `json:"state,omitempty"`
	DisabledComment *DisableComment `json:"disabled_comment,omitempty"`
}

func (r *CreateRequest) Normalize() {
	r.Domain = NormalizeDomainName(r.Domain)
	r.DisabledComment = normalizeDisableComment(r.DisabledComment)
}

func (r CreateRequest) EffectiveState() State {
	if r.State == nil {
		return Active
	}
	return *r.State
}

func (r CreateRequest) Validate() []string {
	fields := make([]string, 0, 3)
	if !IsDomainName(r.Domain) {
		fields = append(fields, "domain")
	}
	if !IsState(r.EffectiveState()) {
		fields = append(fields, "state")
	}
	if !isDisableCommentForState(r.EffectiveState(), r.DisabledComment) {
		fields = append(fields, "disabled_comment")
	}
	return fields
}

type UpdateRequest struct {
	HubSignupDomainID adminspec.HubSignupDomainID `json:"hub_signup_domain_id"`
	Domain            DomainName                  `json:"domain"`
	State             State                       `json:"state"`
	DisabledComment   *DisableComment             `json:"disabled_comment,omitempty"`
}

func (r *UpdateRequest) Normalize() {
	r.Domain = NormalizeDomainName(r.Domain)
	r.DisabledComment = normalizeDisableComment(r.DisabledComment)
}

func (r UpdateRequest) Validate() []string {
	fields := make([]string, 0, 4)
	if !adminspec.IsHubSignupDomainID(r.HubSignupDomainID) {
		fields = append(fields, "hub_signup_domain_id")
	}
	if !IsDomainName(r.Domain) {
		fields = append(fields, "domain")
	}
	if !IsState(r.State) {
		fields = append(fields, "state")
	}
	if !isDisableCommentForState(r.State, r.DisabledComment) {
		fields = append(fields, "disabled_comment")
	}
	return fields
}

func NormalizeDomainName(value DomainName) DomainName {
	normalized := strings.ToLower(strings.TrimSpace(string(value)))
	normalized = strings.TrimSuffix(normalized, ".")
	return DomainName(normalized)
}

func IsDomainName(value DomainName) bool {
	normalized := string(NormalizeDomainName(value))
	if len(normalized) < 3 || len(normalized) > 253 ||
		strings.Count(normalized, ".") < 1 ||
		numericDomainPattern.MatchString(normalized) ||
		!topLevelLetterPattern.MatchString(
			strings.Split(normalized, ".")[strings.Count(normalized, ".")],
		) {
		return false
	}
	for _, label := range strings.Split(normalized, ".") {
		if !domainLabelPattern.MatchString(label) {
			return false
		}
	}
	return true
}

func IsState(value State) bool {
	return value == Active || value == Disabled
}

func IsDisableComment(value DisableComment) bool {
	length := len([]rune(strings.TrimSpace(string(value))))
	return length >= 1 && length <= 500
}

func normalizeDisableComment(value *DisableComment) *DisableComment {
	if value == nil {
		return nil
	}
	normalized := DisableComment(strings.TrimSpace(string(*value)))
	return &normalized
}

func isDisableCommentForState(state State, comment *DisableComment) bool {
	if state == Disabled {
		return comment != nil && IsDisableComment(*comment)
	}
	return comment == nil
}

func isDomainFilterText(value DomainFilterText) bool {
	normalized := strings.ToLower(strings.TrimSpace(string(value)))
	return len(normalized) >= 1 && len(normalized) <= 253 &&
		domainFilterPattern.MatchString(normalized)
}
